### PR TITLE
Enh/Improve email printing

### PR DESCRIPTION
**Pull request by GVodyanov**
Fix https://github.com/nextcloud/mail/issues/8866

TODO:
- [x] Add printing button
- [x] Implement printing iframe
- [x] Add additional information and styling

### CODE REVIEW

**Comment by GVodyanov** [comment]
Current Issue:
image

Longer emails are cut and do not display multiple pages

**Comment by hamza221** [comment]
I think a Print preview would be the easiest way around either a pop up or a new page.
It would also solve the problem of big threads, Theoretically with this approach you would have to programatically  expand all of them one by one and then print. 
Gmail solves this problem by opening a new page with adequate design and triggering the print event instantly

**Comment by GVodyanov** [comment]
> I think a Print preview would be the easiest way around either a pop up or a new page. It would also solve the problem of big threads, Theoretically with this approach you would have to programatically expand all of them one by one and then print. Gmail solves this problem by opening a new page with adequate design and triggering the print event instantly

@hamza221 I found a solution to be overlaying an `iframe` and inserting everything needed into it, tell me if you think it looks sane. I know it's hacky, but I believe anything at this point will be, and doing it like this doesn't require a new view.

**Comment by GVodyanov** [comment]
image

Here is what it currently looks like, and yes, there are multiple pages, they don't get cut off

File: src/components/MenuEnvelope.vue [hunk 1]
```diff
@@ -567,8 +568,45 @@ export default {
...
+			frame.style.position = 'absolute'
+			frame.style.top = '-1000000px'
+			document.body.appendChild(frame)
+			const frameDoc = frame.contentWindow ? frame.contentWindow : frame.contentDocument.document ? frame.contentDocument.document : frame.contentDocument
+			frameDoc.document.open()
+			frameDoc.document.write('<html><head><style>')
+			frameDoc.document.write(printStyles)
+			frameDoc.document.write('</style></head><body>')
+			frameDoc.document.write(document.querySelector('#mail-thread-header h2').outerHTML)
+			frameDoc.document.write(`<h4>${this.envelope.from[0].label}, ${this.envelope.from[0].email}</h4>`)
```
Comment (ChristophWurst) [inline src/components/MenuEnvelope.vue]: you have to sanitize user input before putting it into DOM. the label could contain HTML markup

Comment (GVodyanov) [reply src/components/MenuEnvelope.vue]: I added a method to do that, but tell me if we have a more standard way somewhere

**Comment by ChristophWurst** [comment]
I am not confident with the way this works. Building HTML by hand is scary. Why do we need this separate page? Can't we use the existing page and adjust the layout for print? E.g. do not limit the height of messages in the thread when printing so that everything is included, hide unwanted elements, etc.

**Comment by hamza221** [comment]
> I am not confident with the way this works. Building HTML by hand is scary. Why do we need this separate page? Can't we use the existing page and adjust the layout for print? E.g. do not limit the height of messages in the thread when printing so that everything is included, hide unwanted elements, etc.

We can also use the already existing IFrame but we can't include the sender's address...

**Pull request closed without merging**